File: plugins/native-crypto/ios/Sources/NativeCrypto/NativeCryptoPlugin.swift
import Capacitor
import Foundation

@objc(NativeCryptoPlugin)
public class NativeCryptoPlugin: CAPPlugin {
    private var libraryLoaded = false

    public override func load() {
        super.load()
        libraryLoaded = NativeCryptoLoader.load()
    }

    @objc(init:)
    public func initPlugin(_ call: CAPPluginCall) {
        guard libraryLoaded else {
            call.reject("messie_crypto_ffi native library not loaded")
            return
        }
        call.reject("Native crypto init not implemented yet")
    }

    @objc public func encryptEvent(_ call: CAPPluginCall) {
        call.reject("Native crypto encryptEvent not implemented yet")
    }

    @objc public func decryptEvent(_ call: CAPPluginCall) {
        call.reject("Native crypto decryptEvent not implemented yet")
    }

    @objc public func downloadKeys(_ call: CAPPluginCall) {
        call.reject("Native crypto downloadKeys not implemented yet")
    }

    @objc public func refreshDeviceLists(_ call: CAPPluginCall) {
        call.reject("Native crypto refreshDeviceLists not implemented yet")
    }

    @objc public func getUserVerificationStatus(_ call: CAPPluginCall) {
        call.reject("Native crypto getUserVerificationStatus not implemented yet")
    }

    @objc public func setDeviceVerified(_ call: CAPPluginCall) {
        call.reject("Native crypto setDeviceVerified not implemented yet")
    }

    @objc public func flush(_ call: CAPPluginCall) {
        call.reject("Native crypto flush not implemented yet")
    }

    @objc public func close(_ call: CAPPluginCall) {
        call.reject("Native crypto close not implemented yet")
    }
}

private enum NativeCryptoLoader {
    static func load() -> Bool {
        if libraryLoaded {
            return true
        }
        // Loading the XCFramework happens automatically via linker when bundled.
        libraryLoaded = NSClassFromString("NativeCryptoFFI") != nil
        return libraryLoaded
    }

    private static var libraryLoaded = false
}

File: plugins/native-crypto/android/nativecrypto/src/main/java/com/messie/nativecrypto/NativeCryptoPlugin.kt
package com.messie.nativecrypto

import com.getcapacitor.JSObject
import com.getcapacitor.Logger
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.PluginMethod

@CapacitorPlugin(name = "NativeCrypto")
class NativeCryptoPlugin : Plugin() {
    private var libLoaded: Boolean = false

    override fun load() {
        super.load()
        libLoaded = try {
            System.loadLibrary("messie_crypto_ffi")
            true
        } catch (error: UnsatisfiedLinkError) {
            Logger.warn("NativeCrypto", "messie_crypto_ffi library not present: ${error.message}")
            false
        }
    }

    @PluginMethod
    fun init(call: PluginCall) {
        if (!libLoaded) {
            call.reject("messie_crypto_ffi native library not loaded")
            return
        }
        call.reject("Native crypto init not implemented yet")
    }

    @PluginMethod
    fun encryptEvent(call: PluginCall) {
        call.reject("Native crypto encryptEvent not implemented yet")
    }

    @PluginMethod
    fun decryptEvent(call: PluginCall) {
        call.reject("Native crypto decryptEvent not implemented yet")
    }

    @PluginMethod
    fun downloadKeys(call: PluginCall) {
        call.reject("Native crypto downloadKeys not implemented yet")
    }

    @PluginMethod
    fun refreshDeviceLists(call: PluginCall) {
        call.reject("Native crypto refreshDeviceLists not implemented yet")
    }

    @PluginMethod
    fun getUserVerificationStatus(call: PluginCall) {
        call.reject("Native crypto getUserVerificationStatus not implemented yet")
    }

    @PluginMethod
    fun setDeviceVerified(call: PluginCall) {
        call.reject("Native crypto setDeviceVerified not implemented yet")
    }

    @PluginMethod
    fun flush(call: PluginCall) {
        call.reject("Native crypto flush not implemented yet")
    }

    @PluginMethod
    fun close(call: PluginCall) {
        call.reject("Native crypto close not implemented yet")
    }
}
